rust/parquet_derive/src/parquet_field.rs

```
@@ -0,0 +1,942 @@
+                | Type::Vec(ref second_type)
+                | Type::Array(ref second_type) => match **second_type {
+                    Type::TypePath(_) => Some(self.optional_definition_levels()),
+                    _ => unimplemented!("a little too much nesting. bailing out."),
+                },
+            },
+            Type::Reference(_, ref first_type)
+            | Type::Vec(ref first_type)
+            | Type::Array(ref first_type) => match **first_type {
+                Type::TypePath(_) => None,
+                Type::Reference(_, ref second_type)
+                | Type::Vec(ref second_type)
+                | Type::Array(ref second_type)
+                | Type::Option(ref second_type) => match **second_type {
+                    Type::TypePath(_) => Some(self.optional_definition_levels()),
+                    Type::Reference(_, ref third_type) => match **third_type {
+                        Type::TypePath(_) => Some(self.optional_definition_levels()),
+                        _ => unimplemented!(
+                            "we don't do some more complex definition levels... yet!"
+                        ),
+                    },
+                    _ => unimplemented!(
+                        "we don't do more complex definition levels... yet!"
+                    ),
+                },
+            },
+        };
+
+        let write_batch_expr = if definition_levels.is_some() {
+            quote! {
+                if let #column_writer(ref mut typed) = column_writer {
+                    typed.write_batch(&vals[..], Some(&definition_levels[..]), None).unwrap();
+                } else {
+                    panic!("schema and struct disagree on type for {}", stringify!{#ident})
+                }
+            }
+        } else {
+            quote! {
+                if let #column_writer(ref mut typed) = column_writer {
+                    typed.write_batch(&vals[..], None, None).unwrap();
```

Again, we should surface the potential error from `write_batch` instead of just `unwrap`.